proposals/extensions/android-parcelable.md

```diff
@@ -0,0 +1,254 @@
+
+TODO: Check interactions with inheritance by delegation.
+  
+Some syntactic options:    
+- The names `Parcelize` and `Transient` are just placeholders for now.
+- We could allow the user to omit teh supertype `Parcelable`, to be more DRY and have only the annotation to signify that the class is Parcelable, btu this would be more challenging wrt the tooling support.
+- We could annotate the supertype itself to make it more local: 
+  - `class MyParcelable(val data: Int): @Auto Parcelable`  
+  
+### Generated logic
+
+The generated serialization logic should take into account the following:
+- support for `IBinder`
+- support for collections and arrays
+- support for file descriptors wrt `describeContents()` 
+    
+>Discussion:    
+There is an option to generate two constructors:
+>- one ordinary primary constructor,
+>- another one that takes `Parcel` and creates an object from it.
+>
+>This has a benefit of being more convenient when it comes to hierarchies (see below). The issue here is that the second constructor will have to bypass any initializers in the class body or do sophisticated combination of that logic and the deserialization. Overall it seems easier to have only one traditional constructor and call it from the `createFromParcel` method.  
+    
+### Custom Parcelables
+
+Sometimes users need to write their own custom serialization logic manually, but the boilerplate described above is undesirable. We can simplify this task by introducing the following convention:
+
+```kotlin
+@Parcelize
+class MyParcelable(val data: Int) : Parcelable {
+    companion object : Parceler<MyParcelable> {
+        override fun MyParcelable.writeToParcel(parcel: Parcel, flags: Int) {
+            parcel.writeInt(data)
+        }
+        
+        override fun createFromParcel(parcel: Parcel): MyParcelable {
+            return MyParcelable(parcel.readInt())
+        }
+    }
+}
+```
+
+The `Parceler` interface is defined in the same runtime library that comes with the compiler extension. It's defined as follows:
+
+```kotlin
+interface Parceler<P : Parcelable> {
+    fun describeContents() = 0
+
+    fun P.writeToParcel(parcel: Parcel, flags: Int)
+    
+    fun createFromParcel(parcel: Parcel): P    
+    
+    fun newArray(size: Int): Array<P>
+}
+```
+
+If a `@Parcelize` class has a companion object that implements `Parceler`:
+- the necessary boilerplate is generated for the `Parcelable` convention and delegated to the companion object
+- the `newArray()` method of the companion object is implemented automatically unless it's explicitly overridden
+- the type argument to `Parceler` must match the containing class, otherwise it's a compiler error
+
+Note: Indirect implementations (`object : Foo`, where `Foo : Parceler`) are questionable here, but we can allow them if there are use cases for it.
+
+Syntactic options:
+- It does not have to be a companion object, a named object, e.g. `object Parceler: Parceler<MyPercelable>` may be ok too
```

Comment:
I think it is even _undesirable_ to implement it as companion object, because `MyParcelable.newArray` would be accessible. It's probably better has to be private object.